language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir

```diff
@@ -0,0 +1,125 @@
+//! account: alice
+//! account: vivian, 1000000, 0, validator
+//! account: viola, 1000000, 0, validator
+
+//! new-transaction
+//! sender: alice
+import 0x0.ValidatorConfig;
+
+// register Alice as a validator candidate
+main() {
+    let config: ValidatorConfig.Config;
+
+    ValidatorConfig.register_candidate_validator(h"10", h"20", h"30");
+    return;
+}
+
+// check: EXECUTED
+
+//! new-transaction
+//! sender: alice
+import 0x0.LibraSystem;
+import 0x0.ValidatorConfig;
+
+// rotate alice's pubkey and then run the block prologue. No reconfiguration should be triggered,
+// since alice is not a validator
+// TODO: trigger a BlockMetadata transaction here instead
+main() {
+    ValidatorConfig.rotate_consensus_pubkey(h"40");
+    LibraSystem.block_prologue(1);
+
+    return;
+}
+
+// check: events: []
+// check: EXECUTED
+
+//! new-transaction
+//! sender: vivian
+import 0x0.LibraSystem;
+import 0x0.ValidatorConfig;
+
+// rotate vivian's pubkey and then run the block prologue. Now, reconfiguration should be triggered.
+main() {
+    let info: LibraSystem.ValidatorInfo;
+
+    ValidatorConfig.rotate_consensus_pubkey(h"40");
+    LibraSystem.block_prologue(2);
+    // check that the validator set contains Vivian's new key
+    info = LibraSystem.get_ith_validator_info(1);
+    assert(*LibraSystem.get_consensus_pubkey(&info) == h"40", 99);
+
+    return;
+}
+
+// check: ContractEvent {
+// TODO: check for reconfiguration event type once we have it
+// check: EXECUTED
+
+//! new-transaction
+//! sender: vivian
+import 0x0.LibraSystem;
+import 0x0.ValidatorConfig;
+// rotate vivian's pubkey to the same value and run the block prologue. No reconfiguration should be
```

Comment:
how do we know it didn't happen?

Reply:
Ah: it's subtle, but the `// check: events []` part after the tx is testing this. I'll add a comment to make this explicit.